src/protocol/thread/thread.ts

```diff
@@ -1104,22 +1106,32 @@ class _ThreadFront {
 
   // Get all original/generated IDs which can represent a location in sourceId.
   private _getAlternateSourceIds(sourceId: SourceId) {
+    if (this.alternateSourceIds.has(sourceId)) {
+      return this.alternateSourceIds.get(sourceId)!;
+    }
+
     const rv = new Set<SourceId>();
-    const worklist = [sourceId];
+    let currentSourceId = sourceId;
+    const worklist = [currentSourceId];
+
     while (worklist.length) {
-      sourceId = worklist.pop()!;
-      if (rv.has(sourceId)) {
+      currentSourceId = worklist.pop()!;
+      if (rv.has(currentSourceId)) {
         continue;
       }
-      rv.add(sourceId);
-      const sources = this.sources.get(sourceId);
+      rv.add(currentSourceId);
+      const sources = this.sources.get(currentSourceId);
       assert(sources);
       const { generatedSourceIds } = sources;
-      (generatedSourceIds || []).forEach(id => worklist.push(id));
-      const originalSourceIds = this.originalSources.map.get(sourceId);
-      (originalSourceIds || []).forEach(id => worklist.push(id));
+      worklist.concat(generatedSourceIds || []);
+      const originalSourceIds = this.originalSources.map.get(currentSourceId);
+      worklist.concat(originalSourceIds || []);
```

Comment:
I missed this the first time: `worklist.concat()` will not modify `worklist` but create a new array instead.